### ISSUE TITLE
Add TextTexture for 3D signs

### ISSUE THREAD

**Issue by RobertBColton** [godotengine/godot-proposals#11443]
### Describe the project you are working on

Heavily UI based with a lot of billboard signs with text

### Describe the problem or limitation you are having in your project

I would like to add signs to my game with various text written on them.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

There would be a special texture type called TextTexture that renders text onto it. It would make it so much easier than having to code anything or hack around with viewports. I don't know if MeshTexture is able to do this but I haven't figured out how to use it yet as its not got hardly any documentation. I have however used TextMesh but it's inefficient for simple billboards and especially if you have a lot of text plus it sometimes doesn't work with certain non-monochrome glyphs.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

You want to add a 3D sign to your game, you would:
1) Create a box mesh
2) Add a standard material
3) Assign a new TextTexture in inspector
4) Set the text of the sign as a property on the TextTexture

Voila, 3D sign

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Not really you have to code a viewport and make a mess of your scene since there's no directly renderable textures.

### Is there a reason why this should be core and not an add-on in the asset library?

It's very common feature, in a lot of engines including Unity/Roblox, etc. It's important to a lot of games to just quickly slap some text on 3D signs and stuff.

**Comment by Meorge** [comment]
Is this a feature in Unity? It's been a little while now since I've used the engine, so maybe it's been added since or I've just forgotten about it, but I don't remember there being a built-in way to generate a texture with text on it.

Godot does have [Label3D](https://docs.godotengine.org/en/stable/classes/class_label3d.html), which is pretty similar to a TextMeshPro object in Unity (what you might be thinking of from there).

**Comment by RobertBColton** [comment]
Ok, that's interesting, thank you. That is extremely helpful to me.

However, I am not so sure we can't also still "break up" the implementation. I mean, there could be cases where people would like to put text on other kinds of meshes. Also, it has the advantage of letting you tile the text if needed. Although, I don't think I will personally need it at least for the time being.

**Comment by Calinou** [comment]
If you want actual mesh-based text, you can use a MeshInstance3D node with a TextMesh mesh. This will generally be slower and look worse than Label3D unless you use some kind of antialiasing with subpixel precision (MSAA/TAA/FSR2) though.

> For example, imagine if you wanted the sign bent or curved.

TextMesh can achieve this if you apply a custom shader on a material override set on the MeshInstance3D node. There's an example of this in the [3D Labels and Texts](https://github.com/godotengine/godot-demo-projects/tree/master/3d/labels_and_texts) demo.

**Comment by Meorge** [comment]
For a texture with text displayed on it, that you could apply to any texture-displaying object, it sounds neat but I could also imagine it scope-creeping its way up into something nearly indistinguishable from a separate Viewport.

These are some questions about the functionality I could see people asking:
- Horizontally, can the text be left-aligned, center-aligned, or right-aligned?
- Vertically, can the text be top-aligned, center-aligned, or bottom-aligned?
- Can the text be placed at an arbitrary offset on the texture?
- Can the text be rotated?
- Can we configure how much padding/spacing is there on the sides of the text?
- Can we set the texture's background? Can it be a texture, or a solid color, or transparent?
- Can we set the texture or color of the text?
- Can we give the text a shadow? An outline?
- Can we add other textures (icons and such) within or alongside the text?
- Can these properties/objects within the texture be animated or changed at runtime?

If these things were to be implemented, I feel like it'd more or less just be a second implementation of a Godot SubViewport and ViewportTexture.

**Comment by RobertBColton** [comment]
Yeah so I'm making a city builder in Godot. One of the requirements is I am using only built in shapes to create the art, no use of Blender. I'm rendering small signs to the buildings for their billboards/advertisements. Sometimes the TextMesh will break with certian Unicode Emojis, such as the Radioactive symbol, othertimes it don't. So I'm going to try switching out to Label3D with some extruding to get around that.

![Image](https://github.com/user-attachments/assets/b1c02f13-1ee2-421f-836a-e6c5bad69de1)

**Comment by RobertBColton** [comment]
I'm planning to release the source to the people of Godot.

**Comment by novemberist** [comment]
I think an interesting use case for this proposal would be in conjunction with decals, i.e. quickly projecting some text onto meshes without having to prepare an image texture in an external program.

**Comment by RobertBColton** [comment]
After further consideration of the issue and the original use case, I am deciding to close this proposal. I no longer feel it's a good focus or really needed when it can really be achieved through viewports.

Thank you all for your input and feedback on this one. If anybody sees otherwise, feel free to open a new proposal.
